File: apps/desktop/src/services/commands/analytics.ts
import { invoke } from "@/services/tauri";

import type { DailyProgressPointDto } from "./progress";

export interface MasteryDistributionDto {
  newCount: number;
  learningCount: number;
  reviewingCount: number;
  masteredCount: number;
  total: number;
}

export interface WeakWordDto {
  word: string;
  deckName: string;
  totalReviews: number;
  accuracy: number;
}

export interface DeckBreakdownDto {
  deckName: string;
  wordsReviewed: number;
  accuracy: number;
  totalReviews: number;
}

export interface AnalyticsDto {
  mastery: MasteryDistributionDto;
  weakWords: WeakWordDto[];
  deckBreakdown: DeckBreakdownDto[];
  monthlyActivity: DailyProgressPointDto[];
}

export function getAnalytics(): Promise<AnalyticsDto> {
  return invoke<AnalyticsDto>("get_analytics");
}

File: apps/desktop/src-tauri/src/dto/analytics.rs
use serde::Serialize;

use crate::dto::progress::DailyProgressPointDto;

#[derive(Debug, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MasteryDistributionDto {
    pub new_count: i64,
    pub learning_count: i64,
    pub reviewing_count: i64,
    pub mastered_count: i64,
    pub total: i64,
}

#[derive(Debug, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct WeakWordDto {
    pub word: String,
    pub deck_name: String,
    pub total_reviews: i64,
    pub accuracy: i64,
}

#[derive(Debug, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DeckBreakdownDto {
    pub deck_name: String,
    pub words_reviewed: i64,
    pub accuracy: i64,
    pub total_reviews: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnalyticsDto {
    pub mastery: MasteryDistributionDto,
    pub weak_words: Vec<WeakWordDto>,
    pub deck_breakdown: Vec<DeckBreakdownDto>,
    pub monthly_activity: Vec<DailyProgressPointDto>,
}
